src/pygeosx/pysolver/PySolver.cpp

```diff
@@ -0,0 +1,303 @@
+#define PY_SSIZE_T_CLEAN
+#include <Python.h>
+
+// Source includes
+#include "../pygeosx.hpp"
+#include "PySolver.hpp"
+#include "../PyGroup.hpp"
+
+#define VERIFY_NON_NULL_SELF( self ) \
+  PYTHON_ERROR_IF( self == nullptr, PyExc_RuntimeError, "Passed a nullptr as self.", nullptr )
+
+#define VERIFY_INITIALIZED( self ) \
+  PYTHON_ERROR_IF( self->group == nullptr, PyExc_RuntimeError, "The PySolver is not initialized.", nullptr )
+
+
+namespace geosx
+{
+extern std::unique_ptr< GeosxState > g_state;
+
+extern bool g_alreadyInitialized;
+
+namespace python
+{
+
+struct PySolver
+{
+  PyObject_HEAD
+
+  static constexpr char const * docString =
+    "A Python interface to geosx::SolverBase.";
+
+  geosx::SolverBase * group;
+  geosx::ProblemManager * pb_manager;
+};
+
+static void PySolver_dealloc( PySolver * self )
+{
+  Py_TYPE( self )->tp_free((PyObject *)self );
+}
+
+
+
+static PyObject * PySolver_new( PyTypeObject *type, PyObject *args, PyObject *kwds )
+{
+  GEOSX_UNUSED_VAR( args, kwds );
+  PySolver *self;
+  geosx::ProblemManager * pb_manager = &(g_state->getProblemManager());
+
+  self = (PySolver *)type->tp_alloc( type, 0 );
+  if( self != nullptr )
+  {
+    self->group = nullptr;
+    self->pb_manager = pb_manager;
+
+    if( self->pb_manager == nullptr )
+    {
+      Py_DECREF( self );
+      return nullptr;
+    }
+
+  }
+
+  return (PyObject *)self;
+}
+
+
+
+static int PySolver_init( PySolver *self, PyObject *args, PyObject *kwds )
+{
+  GEOSX_UNUSED_VAR( kwds );
+
+  PYTHON_ERROR_IF( !g_alreadyInitialized, PyExc_RuntimeError, "GEOSX must be initialized first. Call pygeosx.initialize().", -1 );
+
+  geosx::SolverBase *group=nullptr;
+
+  PyObject * unicodePath;
+  if( !PyArg_ParseTuple( args, "U", &unicodePath ) )
+  {
+    return -1;
+  }
+
+  LvArray::python::PyObjectRef<> asciiPath { PyUnicode_AsASCIIString( unicodePath ) };
+  if( asciiPath == nullptr )
+  {
+    return -1;
+  }
+
+  char const * const path = PyBytes_AsString( asciiPath );
+  if( path == nullptr )
+  {
+    return -1;
+  }
+
+  geosx::EventManager & eventManager = self->pb_manager->getEventManager();
+
+  eventManager.forSubGroups< geosx::EventBase >( [&]( geosx::EventBase & subEvent )
+  {
+    subEvent.getTargetReferences();
+  } );
+
+  geosx::EventBase * subEvent = nullptr;
```

Comment:
This looks very similar to `PyGroup::getSolver`. Also not a fan of the two `static_cast`s below.

Reply:
Sorry the getSolver routine was for the second version I had of the PySolver Structure. I decided to not put it in this PR

Reply:
I took the `static_cast` logic directly from the `EventManager::run`

Reply:
Yeah those `static_casts` should be `dynamic_casts`, but this is even more dangerous than that because here you're casting to `SolverBase` instead of `EventBase`.

Reply:
No more `static_cast` with the last commit.